xcm/xcm-builder/src/fungibles_adapter.rs

```diff
@@ -0,0 +1,267 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use sp_std::{prelude::*, result, convert::TryFrom, marker::PhantomData, borrow::Borrow};
+use xcm::v0::{Error as XcmError, Result, MultiAsset, MultiLocation, Junction};
+use frame_support::traits::{Get, tokens::fungibles::Mutate as Fungibles};
+use xcm_executor::traits::{LocationConversion, TransactAsset};
+
+/// Asset transaction errors.
+pub enum Error {
+	/// Asset not found.
+	AssetNotFound,
+	/// `MultiLocation` to `AccountId` conversion failed.
+	AccountIdConversionFailed,
+	/// `u128` amount to currency `Balance` conversion failed.
+	AmountToBalanceConversionFailed,
+	/// `MultiLocation` to `AssetId` conversion failed.
+	AssetIdConversionFailed,
+}
+
+impl From<Error> for XcmError {
+	fn from(e: Error) -> Self {
+		match e {
+			Error::AssetNotFound => XcmError::FailedToTransactAsset("AssetNotFound"),
+			Error::AccountIdConversionFailed =>
+				XcmError::FailedToTransactAsset("AccountIdConversionFailed"),
+			Error::AmountToBalanceConversionFailed =>
+				XcmError::FailedToTransactAsset("AmountToBalanceConversionFailed"),
+			Error::AssetIdConversionFailed =>
+				XcmError::FailedToTransactAsset("AssetIdConversionFailed"),
+		}
+	}
+}
+
+/// Generic third-party conversion trait. Use this when you don't want to force the user to use default
+/// impls of `From` and `Into` for the types you wish to convert between.
+///
+/// One of `convert`/`convert_ref` and `reverse`/`reverse_ref` MUST be implemented. If possible, implement
+/// `convert_ref`, since this will never result in a clone. Use `convert` when you definitely need to consume
+/// the source value.
+pub trait Convert<A: Clone, B: Clone> {
```

Comment:
DQ: What's the difference/motivation relative to `TryFrom` (see docs [here](https://doc.rust-lang.org/std/convert/trait.TryFrom.html))?

Reply:
`TryFrom` is intrinsic to the (source) type in question and as such it is intended for the situation where there is and could only be one sensible way of converting from one to the other type.

In cases where there may be several reasonable ways of converting between two types, then a third-party "converter" type is needed which is able to select/define the means of conversion.